dynet/tensor.cc

```diff
@@ -136,6 +136,23 @@ void TensorTools::CopyElements(const Tensor& v, const Tensor& v_src) {
 #endif
 }
 
+void TensorTools::Clip(Tensor& d, float left, float right){
+#if HAVE_CUDA
```

Comment:
I don't think this is correct on CUDA. It looks like this is declaring a new array of floats completely unrelated to the content of `d`, then clipping it and over-writing `d`.

I think the best way to do this is correctly is creating a functor called `FClip` that performs clipping, then replace this whole function (both CPU and GPU) with:

    d.device(*d.edevice) = d.unaryExpr(FClip(left, right));

Take a look at `functors.h` for example of functors.